resolver/src/name_server_pool.rs

```diff
@@ -201,136 +203,252 @@ impl ConnectionProvider for StandardConnection {
+                socket_addr: config.socket_addr,
+                timeout: options.timeout,
+                tls_dns_name: config.tls_dns_name.clone().unwrap_or_default(),
+            })),
+            #[cfg(feature = "dns-over-https")]
+            Protocol::Https => {
+                ConnectionHandleInner::Connect(Some(ConnectionHandleConnect::Https {
+                    socket_addr: config.socket_addr,
+                    timeout: options.timeout,
+                    tls_dns_name: config.tls_dns_name.clone().unwrap_or_default(),
+                }))
+            }
+            #[cfg(feature = "mdns")]
+            Protocol::Mdns => ConnectionHandleInner::Connect(Some(ConnectionHandleConnect::Mdns {
+                socket_addr: config.socket_addr,
+                timeout: options.timeout,
+            })),
+        };
+
+        ConnectionHandle(Arc::new(Mutex::new(dns_handle)))
+    }
+}
+
+#[derive(Debug)]
+pub enum ConnectionHandleConnect {
+    Udp {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+    },
+    Tcp {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+    },
+    #[cfg(feature = "dns-over-tls")]
+    Tls {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+        tls_dns_name: String,
+    },
+    #[cfg(feature = "dns-over-https")]
+    Https {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+        tls_dns_name: String,
+    },
+    #[cfg(feature = "mdns")]
+    Mdns {
+        socket_addr: SocketAddr,
+        timeout: Duration,
+    },
+}
+
+impl ConnectionHandleConnect {
+    fn connect(self) -> ConnectionHandleConnected {
+        use self::ConnectionHandleConnect::*;
+
+        debug!("connecting: {:?}", self);
+        match self {
+            Udp {
+                socket_addr,
+                timeout,
+            } => {
+                let (stream, handle) = UdpClientStream::new(socket_addr);
                 // TODO: need config for Signer...
                 let dns_conn = DnsMultiplexer::with_timeout(
                     stream,
                     handle,
-                    options.timeout,
+                    timeout,
                     NoopMessageFinalizer::new(),
                 );
 
                 let (stream, handle) = DnsExchange::connect(dns_conn);
-                // TODO: instead of spawning here, return the stream as a "Background" type...
-                tokio::executor::spawn(stream.and_then(|stream| stream).map_err(|e| {
-                    error!("error, udp connection shutting down: {}", e);
-                }));
-
+                let stream = stream.and_then(|stream| stream).map_err(|e| {
+                    debug!("udp connection shutting down: {}", e);
+                });
                 let handle = BufDnsRequestStreamHandle::new(handle);
-                ConnectionHandle::UdpOrTcp(handle)
+
+                tokio::executor::spawn(stream);
```

Comment:
this needs to avoid a panic in global resolver use cases. see: DefaultExecutor in tokio::executor.